src/main/java/org/commcare/formplayer/application/FormSubmissionController.java

```diff
@@ -252,38 +275,100 @@ private SubmitResponseBean processFormXml(FormSubmissionContext context) throws
         return context.success();
     }
 
+    private MultiValueMap<String, HttpEntity<Object>> getMultiPartFormBody(FormSession formSession)
+            throws IOException {
+        MultiValueMap<String, HttpEntity<Object>> body = new LinkedMultiValueMap<>();
+
+        // Add any media files associated with the form
+        Path mediaDirPath = formSession.getMediaDirectoryPath(restoreFactory.getDomain(),
+                restoreFactory.getUsername(), restoreFactory.getAsUsername(), storageFactory.getAppId());
+        File mediaFile = mediaDirPath.toFile();
+        if (mediaFile.exists()) {
+            File[] files = Objects.requireNonNull(mediaDirPath.toFile().listFiles());
+            if (files.length > maxAttachmentsPerForm) {
+                failWithError("form.upload.attachments.threshold.error.message", maxAttachmentsPerForm.toString());
+            }
+            for (File file : files) {
+                validateFile(file);
+
+                String contentType = FileUtils.getContentType(file);
+                if (!StringUtils.hasText(contentType)) {
+                    contentType = "application/octet-stream";
+                }
+
+                HttpEntity<Object> filePart = createFilePart(file.getName(), file.getName(),
+                        Files.readAllBytes(file.toPath()), contentType);
+                body.add(file.getName(), filePart);
+            }
+        }
+
+        // Add the form xml
+        HttpEntity<Object> xmlPart = createFilePart("xml_submission_file", "xml_submission_file.xml",
+                formSession.getInstanceXml(false), "text/xml");
+        body.add("xml_submission_file", xmlPart);
+        return body;
+    }
+
+    private void validateFile(File file) throws FileNotFoundException {
+        if (isUnSupportedFileExtension(file.getName()) && isUnsupportedMimeType(new FileInputStream(file),
+                file.getName())) {
+            failWithError("form.upload.attachment.invalid", file.getName());
+        } else if (isFileTooLarge(file.length())) {
+            failWithError("form.upload.attachment.oversize", file.getName());
+        }
+    }
+
+    private void failWithError(String localeKey, String... args) {
+        String attachmentsThresholdError = Localization.get(localeKey, args);
+        throw new RuntimeException(attachmentsThresholdError);
```

Comment:
Can we use a custom exception class rather than `RuntimeException`. How about `FormAttachmentException`?